File: C /2966. Divide Array Into Arrays With Max Difference.c
// Comparison function for qsort
int cmp(const void *a, const void *b) {
    return *(int *)a - *(int *)b;
}

// Function to divide array into subarrays of size 3
int** divideArray(int* nums, int numsSize, int k, int* returnSize, int** returnColumnSizes) {
    // Sort the array in ascending order
    qsort(nums, numsSize, sizeof(int), cmp);

    // Calculate the number of subarrays needed
    *returnSize = numsSize / 3;

    // Allocate memory for storing subarrays and their sizes
    int **arr = (int **)malloc(sizeof(int *) * (*returnSize));
    *returnColumnSizes = (int *)malloc(sizeof(int) * (*returnSize));

    // Iterate through the sorted array and create subarrays
    for (int i = 0, arrIdx = 0; i < numsSize; i += 3) {
        // Check if the conditions are satisfied
        if (nums[i + 2] - nums[i] > k) {
            *returnSize = 0; // Set returnSize to 0 indicating failure
            return NULL;
        } else {
            // Store the subarray and its size
            (*returnColumnSizes)[arrIdx] = 3;
            arr[arrIdx] = (int *)malloc(sizeof(int) * 3);
            arr[arrIdx][0] = nums[i];
            arr[arrIdx][1] = nums[i + 1];
            arr[arrIdx][2] = nums[i + 2];
            arrIdx++;
        }
    }

    return arr;
}

File: Leetcode/2966. Divide Array Into Arrays With Max Difference.java
// import java.util.Arrays;

public class Solution {
    
    // Method to divide the array into subarrays
    public static int[][] divideArray(int[] nums, int k) {
        // Check if nums is null or empty
        if (nums == null || nums.length == 0) {
            return null;
        }

        // Sort the array in ascending order
        Arrays.sort(nums);
        
        // Calculate the number of subarrays needed
        int numSubarrays = nums.length / 3;

        // Initialize the 2D array to store subarrays
        int[][] output = new int[numSubarrays][3];

        // Populate the 2D array with subarrays
        int m = 0;
        for (int i = 0; i < numSubarrays; i++) {
            for (int j = 0; j < 3; j++) {
                output[i][j] = nums[m];
                m++;
            }
        }

        // Check if the constraint is satisfied for each subarray
        for (int i = 0; i < numSubarrays; i++) {
            if (output[i][2] - output[i][0] > k) {
                // Constraint not satisfied, return an empty array
                return new int[0][];
            }
        }

        // Return the populated 2D array of subarrays
        return output;
    }
}
